File: GuardianMonsters/ui/src/de/limbusdev/guardianmonsters/ui/widgets/ItemListWidget.java
package de.limbusdev.guardianmonsters.ui.widgets;

import com.badlogic.ashley.signals.Listener;
import com.badlogic.ashley.signals.Signal;
import com.badlogic.gdx.scenes.scene2d.Group;
import com.badlogic.gdx.scenes.scene2d.InputEvent;
import com.badlogic.gdx.scenes.scene2d.ui.Button;
import com.badlogic.gdx.scenes.scene2d.ui.ButtonGroup;
import com.badlogic.gdx.scenes.scene2d.ui.ScrollPane;
import com.badlogic.gdx.scenes.scene2d.ui.Skin;
import com.badlogic.gdx.scenes.scene2d.ui.Table;
import com.badlogic.gdx.scenes.scene2d.ui.TextButton;
import com.badlogic.gdx.utils.Align;
import com.badlogic.gdx.utils.Array;
import com.badlogic.gdx.utils.ArrayMap;

import de.limbusdev.guardianmonsters.guardians.items.Inventory;
import de.limbusdev.guardianmonsters.guardians.items.Item;
import de.limbusdev.guardianmonsters.guardians.items.ItemSignal;
import de.limbusdev.guardianmonsters.services.Services;

/**
 * @author Georg Eckert 2017
 */

public class ItemListWidget extends Group implements Listener<ItemSignal> {

    private Table itemTable;
    private Inventory inventory;
    private Skin skin;
    private int lastChosenItem=0;
    private ClickListener clickListener;
    private Array<Item.Category> currentFilters;

    public interface ClickListener {

        void onChoosingItem(Item item);
    }

    public ItemListWidget(Inventory inventory, ClickListener handler, Array<Item.Category> filters)
    {
        this.inventory = inventory;
        this.skin = Services.UI().getInventorySkin();
        this.clickListener = handler;
        this.currentFilters = filters;

        itemTable = new Table();
        itemTable.align(Align.topLeft);

        ScrollPane scrollPane = new ScrollPane(itemTable, skin);


        scrollPane.setSize(192,200);
        scrollPane.setPosition(2,2);
        scrollPane.setScrollBarPositions(false,true);
        addActor(scrollPane);

        init(inventory, filters);
    }

    private void init(Inventory inventory, Array<Item.Category> filters)
    {
        itemTable.clearChildren();
        this.inventory = inventory;
        inventory.add(this);

        Array<Button> buttons = new Array<>();
        final ButtonGroup<TextButton> btnGroup = new ButtonGroup<>();
        btnGroup.setMinCheckCount(1);
        btnGroup.setMaxCheckCount(1);

        // Only proceed if there are any items
        if(inventory.getItems().size > 0) {

            ArrayMap<Integer, Item> filteredItems = new ArrayMap<>();
            int counter = 0;
            for (final Item i : inventory.getItems().keys()) {

                if(filters.contains(Item.Category.ALL, false) || filters.contains(i.getCategory(), false)) {

                    final ItemInventoryButton item = new ItemInventoryButton(i, skin, "item-button-sandstone", inventory);

                    itemTable.add(item).width(192)                                                                                                                                                                                                                                                                                                      .height(40);
                    buttons.add(item);
                    btnGroup.add(item);
                    itemTable.row().spaceBottom(1);

                    item.addListener(new SimpleClickListener(() -> {

                        clickListener.onChoosingItem(i);
                        lastChosenItem = btnGroup.getCheckedIndex();
                    }));

                    if (counter == 0) {
                        item.setChecked(true);
                        clickListener.onChoosingItem(i);
                        inventory.removeAllListeners();
                        inventory.add(this);
                    }
                    filteredItems.put(counter, i);
                    counter++;
                }
            }

            // set to item next to that one, which got deleted
            if(btnGroup.getButtons().size > 0) {

                if (lastChosenItem > btnGroup.getButtons().size - 1) {
                    lastChosenItem = btnGroup.getButtons().size - 1;
                }

                btnGroup.setChecked(btnGroup.getButtons().get(lastChosenItem).getText().toString());
                clickListener.onChoosingItem(filteredItems.get(lastChosenItem));
            }

        }
    }

    public void applyFilter(Item.Category filter)
    {
        Array<Item.Category> filters = new Array<>();
        filters.add(filter);
        applyFilter(filters);
    }

    public void applyFilter(Array<Item.Category> filters) {

        lastChosenItem = 0;
        currentFilters = filters;
        init(inventory, filters);
    }

    @Override
    public void receive(Signal<ItemSignal> signal, ItemSignal itemSignal) {

        if(itemSignal.getMessage() == ItemSignal.Message.DELETED) {
            // Refresh Inventory List, if last item of it's kind got removed
            if (!inventory.containsItem(itemSignal.getItem())) {
                init(inventory, currentFilters);
            }
        }
    }
}

File: GuardianMonsters/ui/src/de/limbusdev/guardianmonsters/ui/widgets/ItemListWidget.kt
package de.limbusdev.guardianmonsters.ui.widgets

import com.badlogic.ashley.signals.Listener
import com.badlogic.ashley.signals.Signal
import com.badlogic.gdx.scenes.scene2d.Group
import com.badlogic.gdx.scenes.scene2d.ui.Button
import com.badlogic.gdx.scenes.scene2d.ui.ButtonGroup
import com.badlogic.gdx.scenes.scene2d.ui.Skin
import com.badlogic.gdx.scenes.scene2d.ui.Table
import com.badlogic.gdx.scenes.scene2d.ui.TextButton
import com.badlogic.gdx.utils.Align
import com.badlogic.gdx.utils.Array
import com.badlogic.gdx.utils.ArrayMap

import de.limbusdev.guardianmonsters.guardians.items.Inventory
import de.limbusdev.guardianmonsters.guardians.items.Item
import de.limbusdev.guardianmonsters.guardians.items.ItemSignal
import de.limbusdev.guardianmonsters.scene2d.Scene2DLayout
import de.limbusdev.guardianmonsters.scene2d.makeScrollPane
import de.limbusdev.guardianmonsters.scene2d.replaceOnButtonClick
import de.limbusdev.guardianmonsters.services.Services

/**
 * ItemListWidget displays a list of several items.
 *
 * @author Georg Eckert 2017
 */

class ItemListWidget
(
        private var inventory: Inventory,
        private val onItemButton: (Item) -> Unit,
        private var currentFilters: Array<Item.Category>,
        private val skin: Skin = Services.UI().inventorySkin
)
    : Group(), Listener<ItemSignal>
{
    // --------------------------------------------------------------------------------------------- PROPERTIES
    private val itemTable = Table()
    private var lastChosenItem = 0


    // --------------------------------------------------------------------------------------------- CONSTRUCTORS
    init
    {
        itemTable.align(Align.topLeft)

        makeScrollPane(itemTable, skin, Scene2DLayout(192f, 200f, 2f, 2f), false, true, this)
    }

    fun initialize(inventory: Inventory, filters: Array<Item.Category>)
    {
        itemTable.clearChildren()
        this.inventory = inventory
        inventory.add(this)

        val buttons = Array<Button>()
        val buttonGroup = ButtonGroup<TextButton>()
        buttonGroup.setMinCheckCount(1)
        buttonGroup.setMaxCheckCount(1)

        // Only proceed if there are any items
        if (inventory.items.size > 0)
        {
            val filteredItems = ArrayMap<Int, Item>()
            var counter = 0
            for (item in inventory.items.keys())
            {
                if (filters.contains(Item.Category.ALL, false) || filters.contains(item.category, false))
                {
                    val itemButton = ItemInventoryButton(item, skin, "item-button-sandstone", inventory)

                    itemTable.add(itemButton).width(192f).height(40f)
                    buttons.add(itemButton)
                    buttonGroup.add(itemButton)
                    itemTable.row().spaceBottom(1f)

                    itemButton.replaceOnButtonClick {

                        onItemButton(item)
                        lastChosenItem = buttonGroup.checkedIndex
                    }

                    if (counter == 0) {
                        itemButton.isChecked = true
                        onItemButton(item)
                        inventory.removeAllListeners()
                        inventory.add(this)
                    }
                    filteredItems.put(counter, item)
                    counter++
                }
            }

            // set to item next to that one, which got deleted
            if (buttonGroup.buttons.size > 0) {

                if (lastChosenItem > buttonGroup.buttons.size - 1) {
                    lastChosenItem = buttonGroup.buttons.size - 1
                }

                buttonGroup.setChecked(buttonGroup.buttons.get(lastChosenItem).text.toString())
                onItemButton(filteredItems.get(lastChosenItem))
            }

        }
    }


    // --------------------------------------------------------------------------------------------- METHODS
    fun applyFilter(filter: Item.Category) {
        val filters = Array<Item.Category>()
        filters.add(filter)
        applyFilter(filters)
    }

    fun applyFilter(filters: Array<Item.Category>) {

        lastChosenItem = 0
        currentFilters = filters
        initialize(inventory, filters)
    }

    override fun receive(signal: Signal<ItemSignal>, itemSignal: ItemSignal) {
        initialize(inventory, currentFilters)
    }
}
